test/controllers/application_controller_test.rb

```diff
@@ -7,15 +7,32 @@ class ApplicationControllerTest < ActionDispatch::IntegrationTest
     @admin_user = create(:admin_user)
   end
 
-  test "should redirect to member loans after login" do
+  test "should redirect to member loans after login if there is no referer" do
     sign_in(@user)
     post user_session_url
     assert_redirected_to member_loans_path
   end
 
-  test "should redirect to admin dashboard after login" do
+  test "should redirect to admin dashboard after login if there is no referer" do
     sign_in(@admin_user)
     post user_session_url
     assert_redirected_to admin_dashboard_path
   end
+
+  test "should redirect to previous page for admin after login if there is a referer" do
+    get new_user_session_path, headers: {referer: "http://www.example.com/original-page"}
+    post user_session_path, params: {user: {email: @admin_user.email, password: "password"}}
```

Comment:
I think part of the issue was that using the `sign_in` helper from Devise circumvents the code in `Users::SessionsController#new`, so that the referer, even if it was set correctly, wasn't getting checked. This test avoids that by making those requests manually.